data/provider/src/main/java/tech/pegasys/teku/api/blockrootselector/BlockRootSelectorFactory.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Consensys Software Inc., 2023
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.api.blockrootselector;
+
+import static tech.pegasys.teku.spec.config.SpecConfig.GENESIS_SLOT;
+
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.api.AbstractSelectorFactory;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.metadata.ObjectAndMetaData;
+import tech.pegasys.teku.storage.client.ChainHead;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
+
+public class BlockRootSelectorFactory extends AbstractSelectorFactory<BlockRootSelector> {
+
+  private final Spec spec;
+
+  public BlockRootSelectorFactory(final Spec spec, CombinedChainDataClient client) {
+    super(client);
+    this.spec = spec;
+  }
+
+  @Override
+  public BlockRootSelector blockRootSelector(final Bytes32 blockRoot) {
+    return () -> client.getBlockByBlockRoot(blockRoot).thenApply(this::fromBlock);
+  }
+
+  @Override
+  public BlockRootSelector headSelector() {
+    return () ->
+        client
+            .getChainHead()
+            .map(this::fromChainHead)
+            .orElse(SafeFuture.completedFuture(Optional.empty()));
+  }
+
+  @Override
+  public BlockRootSelector finalizedSelector() {
+    return () ->
+        SafeFuture.completedFuture(
+            lookupBlockRootData(client.getFinalizedBlock(), client.isChainHeadOptimistic()));
```

Comment:
we may be able to be more tricky here...
Rather than getting the finalized block, theres a finalized Checkpoint on recentChainData, so we might be able to make a combined client fn that does something like 
```
recentChainData.getFinalizedCheckpoint().orElseThrow().getRoot()
```
will be worth investigating

Reply:
it looks like you've done the fn below just not called it...